enhancements/cvo/upgrade_preflight_checks.md

```diff
@@ -0,0 +1,107 @@
+---
+title: upgrade-preflight-checks-for-operators
+authors:
+  - "@LalatenduMohanty"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-06-04
+last-updated: 2020-06-05
+status: implementable
+---
+
+# Upgrade Preflight Checks For Operators
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+* How CVO can get the list of operators which has the preflight checks implemented?
+* What information CVO needs from the operators to create the preflight jobs?
+* Does CVO already has capability to create jobs for preflight checks?
+* How to trigger execution of preflight checks?
+* How CVO will report when it gets the result from preflight checks?
+* What information CVO needs from operators to create the preflight job?
+* How will preflight checks report back success or failure, and the detailed info for administrators?
+
+## Summary
+
+The goal of over the air update(OTA) feature of OpenShift is to make upgrades seamless and make the upgrade a success almost always.
+To improve the upgrade success we should be able to detect if the new version is compatible with the existing version before running the actual upgrade.
+As OpenShift V4 is composed of many operators, we need to give operators ability to run some checks to find out if the new version is compatible with the existing version and use the result to improve the success rate of upgrades.
+
+## Motivation
+
+CVO managed operators should be able to run preflight checks before the actual upgrade.
+It would help operators to find out if they can move to the specific version safely before doing an actual upgrade.
+Also administrators are empowered to run these preflight checks without doing an upgrade.
```

Comment:
Probably distinguish vs. the preconditions from openshift/cluster-version-operator#243 with something like:

> The CVO already supports preconditions including checks for ClusterOperators with Updgradeable=False conditions.  Those checks require the initial version to contain logic for "is the target I'm updating to compatible?".  That works when we can guess at build-time what cluster details may be incompatible with future versions, but does not allow for newly discovered issues to block updates.  Preflight checks, on the other hand, are compiled into the target version, asking "is the source from which I'd be updating compatible?".

Reply:
Changed the wording in https://github.com/openshift/enhancements/pull/363/commits/54725582ca3cd85d78ec02ed237ba72fbad27efd